src/util/mpir_pmi.c

```diff
@@ -187,7 +187,7 @@ int MPIR_pmi_kvs_put(const char *key, const char *val)
 #elif defined(USE_PMIX_API)
     pmix_value_t value;
     value.type = PMIX_STRING;
-    value.data.string = val;
+    value.data.string = (char *) val;
```

Comment:
These changes to `MPIR_pmi_kvs_put` and `MPIR_pmi_barrier_local` seem to be unrelated to the bcast and allgather implementations.

Reply:
The fix to MPIR_pmi_kvs_put is a bug fix (for `PMIx`). 

`MPIR_pmi_barrier_local` originally was only used in `MPIDU_shm_seg_alloc` (that is the reason I refactor it out), but later turns out we don't really need it. `PMI1/2` don't support barrier local anyway. `PMIx` will block and wait for the key anyway. There are also some `MPIR_pmi_` functions in the end may not be needed, e.g. seems we only need `MPIR_pmi_allgather` for the node roots. So I am thinking leaving the interface pruning to a later PR when code stabilizes.

I could split the PR is you insist.

Reply:
I am reviewing the changes in this commit. The commit message says `pmi: implement MPIR_pmi_bcast and MPIR_pmi_allgather`, yet it is changing this functions and I do not understand why. This either should be explained in the commit message, or they should be moved to another commit explaining why they are being changed. They do not need to move to another PR, but they do not appear to belong in this commit.

Reply:
I see. It is a mix-up during warnings fix. I'll correct it.

Reply:
fixed.